base/src/main/java/org/aya/core/visitor/Expander.java

```diff
@@ -0,0 +1,171 @@
+        var newBody = matchy.body().rename().subst(subst).lift(ulift);
+        return Option.some(new WithPos<>(matchy.sourcePos(), newBody));
+      } else if (!orderIndependent && termSubst.getErr()) return Option.none();
+    }
+    return Option.none();
+  }
+  default Option<WithPos<Term>> tryUnfoldClauses(
+    boolean orderIndependent, SeqLike<Arg<Term>> args,
+    int ulift, ImmutableSeq<Matching> clauses
+  ) {
+    return tryUnfoldClauses(orderIndependent, args, new Subst(MutableMap.create()), ulift, clauses);
+  }
+
+  record Normalizer(TyckState state) implements Expander {
+    @Override public Term post(Term term) {
+      return switch (term) {
+        case ElimTerm.App app && app.of() instanceof IntroTerm.Lambda lam -> apply(CallTerm.make(lam, app.arg()));
+        case ElimTerm.App app -> CallTerm.make(app.of(), app.arg());
+        case ElimTerm.Proj proj && proj.of() instanceof IntroTerm.Tuple tup -> {
+          var ix = proj.ix();
+          assert tup.items().sizeGreaterThanOrEquals(ix) && ix > 0 : proj.toDoc(DistillerOptions.debug()).debugRender();
+          yield apply(tup.items().get(ix - 1));
+        }
+        default -> Expander.super.post(term);
+      };
+    }
+  }
+
+  record WHNFer(TyckState state) implements Expander {
+    @Override public Term post(Term term) {
+      return switch (term) {
+        case ElimTerm.App app && app.of() instanceof IntroTerm.Lambda lambda ->
+          apply(CallTerm.make(lambda, app.arg()));
+        case ElimTerm.Proj proj && proj.of() instanceof IntroTerm.Tuple tup -> {
+          var ix = proj.ix();
+          assert tup.items().sizeGreaterThanOrEquals(ix) && ix > 0 : proj.toDoc(DistillerOptions.debug()).debugRender();
+          yield apply(tup.items().get(ix - 1));
+        }
+        default -> Expander.super.post(term);
+      };
+    }
+
+    @Override public Term apply(Term term) {
+      return switch (term) {
+        case IntroTerm.Lambda lambda -> lambda;
+        case IntroTerm.Tuple tuple -> tuple;
+        case FormTerm.Pi pi -> pi;
+        case FormTerm.Sigma sigma -> sigma;
+        case CallTerm.Data data -> data;
+        default -> Expander.super.apply(term);
+      };
+    }
+  }
+
+  record Tracked(
+    @NotNull Set<@NotNull Var> unfolding,
+    @NotNull MutableSet<@NotNull Var> unfolded,
+    @Nullable TyckState state,
+    @NotNull PrimDef.Factory factory
+  ) implements Expander {
+    @Override public Term apply(Term term) {
+      return switch (term) {
+        case CallTerm.Fn fn -> {
+          if (!unfolding.contains(fn.ref())) yield fn;
+          unfolded.add(fn.ref());
+          yield Expander.super.apply(fn);
+        }
+        case CallTerm.Con con -> {
+          if (!unfolding.contains(con.ref())) yield con;
+          unfolded.add(con.ref());
+          yield Expander.super.apply(con);
+        }
+        case CallTerm.Prim prim -> {
+          // TODO[kiva]: Q: is OK to use `state`? so we don't need this override.
```

Comment:
@imkiva

Reply:
Changing the `state` to `@NotNull` resolves this TODO.

Reply:
But @wsx-ucb said some API passes null to it

Reply:
Can you check them for me? @imkiva

Reply:
![image](https://user-images.githubusercontent.com/9902643/183583810-af577b07-5693-4d26-967f-35004091180e.png)

Reply:
Even in this picture, there's a `var state = new TyckState(res._1);` which should be passed to normalize

Reply:
@wsx-ucb can you fix it for us? 😉

Reply:
I guess @wsx-ucb was confused by `PrimDef.Factory.unfold` which takes a `@Nullable TyckState` and then passed it to the normalizer. But the `unfold` was always called by the normalizer with a not-null tyck state. We should make them `@NotNull` altogether.


![image](https://user-images.githubusercontent.com/9902643/183585029-18ba7498-13d2-4afd-a319-8c7cd3510673.png)

Reply:
I think I'm gonna do ti

Reply:
> I think I'm gonna do ti

btw there's a duplicate but unused normalizer in `EndoFunctor`, you can also delete it